Kernel/Storage/SD/SDHostController.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2023, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Function.h>
+#include <AK/Result.h>
+#include <AK/Types.h>
+#include <Kernel/Locking/Mutex.h>
+#include <Kernel/Storage/SD/Commands.h>
+#include <Kernel/Storage/SD/Registers.h>
+#include <Kernel/Storage/SD/SDMemoryCard.h>
+
+namespace Kernel {
+
+class SDHostController : public StorageController {
+    friend class SDMemoryCard;
+
+public:
+    static ErrorOr<NonnullRefPtr<SDHostController>> try_initialize();
+
+    virtual ~SDHostController() = default;
+
+    virtual LockRefPtr<StorageDevice> device(u32 index) const override { return index == 0 ? m_card : nullptr; }
+    virtual ErrorOr<void> reset() override;
+    virtual ErrorOr<void> shutdown() override;
+    virtual size_t devices_count() const override { return m_card ? 1 : 0; }
+    virtual void complete_current_request(AsyncDeviceRequest::RequestResult) override;
+
+protected:
+    virtual SD::SDRegisters volatile* get_register_map_base_address() = 0;
+
+    SDHostController(u32 hardware_relative_controller_id);
+
+private:
+    ErrorOr<void> initialize();
+    ErrorOr<NonnullLockRefPtr<SDMemoryCard>> try_initialize_inserted_card();
```

Comment:
Do we need the SDMemoryCard to be of type `NonNullLockRefPtr` instead of a plain `NonNullRefPtr`? From what I see, after Init, you don't touch the m_card at all.
I think there is a general effort to reduce the usage of `NonNullLockRefPtr`.

Reply:
Right now `SDMemoryCard` does not need to be in a `LockRefPtr`, but there's an overridden method `device` in `SDHostController` that must return a `LockRefPtr`.

Is it safe to store `SDMemoryCard` in a normal `RefPtr` and then have `device` look something like this, which converts from `SDMemoryCard*` to `LockRefPtr<SDMemoryCard>`?

```cpp
virtual LockRefPtr<StorageDevice> device(u32 index) const override { return index == 0 ? m_card.ptr() : nullptr; }
```

Reply:
You are right. I think you can leave it as is. We probably need to change how we manage `StorageDevice` in Storagemanagement, then we can change this to return RefPtr instead of LockRefPtr.